Nama        : Tiur Juliana Sitorus
NIM         : 42324048
Prodi       : DIII Teknologi Informasi
Mata Kuliah : Kecerdasan Buatan

In [5]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
# Load dataset Iris dari scikit-learn
iris = load_iris()

Penjelasan :
Kode ini memanggil library yang dibutuhkan. NumPy dipakai untuk mengolah angka dan Pandas untuk mengolah data berbentuk tabel. Setelah itu, dataset Iris diambil dari Scikit-Learn dengan load_iris() dan disimpan ke variabel iris.

In [6]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]
# Definisikan X (fitur) dan y (target/label)
X = df.drop(columns=["Species"])
y = df["Species"]
print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


Penjelasan :
Data Iris diubah menjadi tabel (DataFrame), lalu ditambahkan kolom Species berisi nama spesies bunga. X berisi empat ukuran bunga sebagai fitur, sedangkan y berisi nama spesies sebagai target yang akan ditebak. print(X) menampilkan tabel fitur berukuran 150 baris dan 4 kolom.

In [7]:
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

Penjelasan :
Kode ini menampilkan isi y, yaitu label spesies setiap bunga. Labelnya masih berupa teks (setosa, versicolor, dan virginica), masing-masing sebanyak 50 data.

In [8]:
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

Penjelasan :
LabelEncoder mengubah label teks menjadi angka supaya bisa diproses model: setosa menjadi 0, versicolor menjadi 1, dan virginica menjadi 2.

In [9]:
from sklearn.model_selection import train_test_split

X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

Penjelasan :
Data dibagi menjadi 80% data latih (120 data) dan 20% data uji (30 data). random_state=0 membuat pembagiannya selalu sama, sehingga setiap fitur tetap berpasangan dengan labelnya.

In [10]:
from sklearn.naive_bayes import GaussianNB

model = GaussianNB()
NB_model = model.fit(X_train, y_train)

Penjelasan :
Model yang dipakai adalah Gaussian Naïve Bayes, karena semua fitur Iris berupa angka kontinu (bisa bernilai desimal). Saat dilatih dengan fit(), model menghitung rata-rata dan standar deviasi setiap fitur untuk setiap spesies. Nilai-nilai ini nanti dipakai untuk menghitung peluang sebuah bunga termasuk spesies tertentu.

In [11]:
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

Penjelasan :
Model menebak spesies dari 30 data uji. Hasilnya berupa angka 0, 1, dan 2 yang mewakili setosa, versicolor, dan virginica.

In [12]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


Penjelasan :
Akurasi dihitung dengan membandingkan hasil tebakan dan label asli, lalu dikali 100 supaya menjadi persen. Hasilnya 96.67%, artinya 29 dari 30 data uji ditebak dengan benar.

In [13]:
#prediksi inputan data baru
#input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


c:\Users\tiurs\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


Penjelasan :
Kode ini menebak spesies untuk satu bunga baru, lalu inverse_transform() mengubah hasil angkanya kembali menjadi nama spesies. Hasilnya setosa. Di output akan muncul tulisan merah UserWarning. Ini bukan error, hanya peringatan. Penyebabnya, model dilatih memakai DataFrame yang punya nama kolom, sedangkan input di sini berupa list biasa tanpa nama kolom.

In [14]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)

result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


Penjelasan :
Ini adalah cara yang benar. Input dibungkus menjadi DataFrame dengan nama kolom yang sama seperti data latih, sehingga warning tidak muncul lagi. Hasilnya tetap setosa.

Tugas
1. Ikuti semua tahapan dalam modul ini.

2. Prediksi data baru masuk ke species apa: [7.3, 3.8, 4.6, 2.3]

In [15]:
input_data = pd.DataFrame(np.array([[7.3, 3.8, 4.6, 2.3]]), columns=iris.feature_names)

result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['virginica']


Penjelasan :
Di tugas nomor 2 ini kami diminta menebak bunga dengan ukuran [7.3, 3.8, 4.6, 2.3] masuk ke spesies apa. Untuk menebaknya, kami memakai model Gaussian Naïve Bayes yang sudah dilatih di tahap sebelumnya.
Pertama, data bunga baru kami ubah dulu menjadi DataFrame dan kami beri nama kolom yang sama dengan data latih. Kalau datanya langsung dimasukkan dalam bentuk list biasa, hasilnya tetap keluar, tapi muncul warning seperti di In [9] modul. Supaya bersih, kami ikuti cara di In [10].
Setelah itu data dimasukkan ke NB_model.predict(). Di tahap ini model membandingkan ukuran bunga baru dengan ciri-ciri tiap spesies yang sudah dipelajarinya waktu training, lalu menghitung peluang bunga itu masuk ke setosa, versicolor, atau virginica. Spesies dengan peluang paling besar dipilih sebagai hasil. Karena hasilnya masih berupa angka, kami ubah lagi menjadi nama spesies pakai inverse_transform(), lalu kami tampilkan dengan print().

Hasilnya bunga tersebut masuk ke spesies virginica. Menurut kami hasil ini cukup masuk akal. Sepal length-nya 7.3 dan petal width-nya 2.3, dua ukuran yang tergolong besar, dan virginica memang spesies dengan ukuran bunga paling besar dibanding dua spesies lainnya. Petal length 4.6 sebenarnya lebih dekat ke versicolor, tapi karena fitur yang lain lebih condong ke virginica, model tetap memilih virginica.
Dari sini kami jadi lebih paham bahwa Naïve Bayes tidak hanya melihat satu fitur saja. Model menggabungkan semua fitur untuk menentukan hasil akhirnya.